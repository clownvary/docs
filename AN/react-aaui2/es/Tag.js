import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent } from 'react';
import { string, func, bool } from 'prop-types';
import classNames from 'classnames';

import { omit, noop } from './shared/utils';
import KEY_CODES from './shared/keyCodes';

var Tag = function (_PureComponent) {
  _inherits(Tag, _PureComponent);

  function Tag(props) {
    _classCallCheck(this, Tag);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props));

    _this.setInputRef = function (input) {
      _this.input = input;
    };

    _this.toInput = function (e) {
      e && e.stopPropagation && e.stopPropagation();

      _this.setState({
        isInput: true,
        isInputCancel: false
      }, function () {
        var input = _this.input;

        if (input) {
          input.focus();
          input.select();
        }
      });
    };

    _this.handleKeyDown = function (e) {
      var target = e.target;

      switch (e.keyCode) {
        case KEY_CODES.ESCAPE:
          e.preventDefault();
          _this.setState({
            isInputCancel: true,
            isTabEnter: false
          }, function () {
            target.blur();
          });
          break;
        case KEY_CODES.TAB:
        case KEY_CODES.ENTER:
          e.preventDefault();

          _this.setState({
            isTabEnter: true
          }, function () {
            target && target.blur();
          });
          break;
        default:
          break;
      }
    };

    _this.handleBlur = function (e) {
      var value = e.target.value;

      _this.setState({
        isInput: false
      }, function () {
        if (!_this.state.isInputCancel) {
          _this.props.onChange(value, _this.state.isTabEnter);
        } else {
          _this.props.onCancel();
        }
      });
    };

    _this.state = {
      isInput: false,
      isTabEnter: false
    };
    return _this;
  }

  Tag.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        editMode = _props.editMode,
        onClose = _props.onClose,
        value = _props.value,
        isNew = _props.isNew,
        rest = _objectWithoutProperties(_props, ['className', 'editMode', 'onClose', 'value', 'isNew']);

    var isInput = this.state.isInput;

    var classes = classNames('tag-input', className);

    if (!editMode) {
      return React.createElement(
        'span',
        { className: 'badge' },
        value
      );
    }

    return React.createElement(
      'span',
      _extends({ className: classes }, omit(rest, ['onChange', 'onCancel'])),
      isInput ? React.createElement('input', {
        className: 'input',
        ref: this.setInputRef,
        placeholder: isNew ? value : undefined,
        defaultValue: isNew ? undefined : value,
        onKeyDown: this.handleKeyDown,
        onBlur: this.handleBlur
      }) : React.createElement(
        'span',
        { className: 'badge badge--with-close-icon' },
        React.createElement(
          'span',
          { onClick: this.toInput },
          value
        ),
        React.createElement('i', { className: 'icon-close-thin', onClick: onClose })
      )
    );
  };

  return Tag;
}(PureComponent);

Tag.displayName = 'AUITag';
Tag.propTypes = {
  isNew: bool,
  editMode: bool,
  className: string,
  value: string,
  onChange: func,
  onCancel: func,
  onClose: func
};
Tag.defaultProps = {
  editMode: false,
  onChange: noop,
  onCancel: noop,
  onClose: noop
};
export default Tag;