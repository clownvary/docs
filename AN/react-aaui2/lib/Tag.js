'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _utils = require('./shared/utils');

var _keyCodes = require('./shared/keyCodes');

var _keyCodes2 = _interopRequireDefault(_keyCodes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Tag = function (_PureComponent) {
  (0, _inherits3.default)(Tag, _PureComponent);

  function Tag(props) {
    (0, _classCallCheck3.default)(this, Tag);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props));

    _this.setInputRef = function (input) {
      _this.input = input;
    };

    _this.toInput = function (e) {
      e && e.stopPropagation && e.stopPropagation();

      _this.setState({
        isInput: true,
        isInputCancel: false
      }, function () {
        var input = _this.input;

        if (input) {
          input.focus();
          input.select();
        }
      });
    };

    _this.handleKeyDown = function (e) {
      var target = e.target;

      switch (e.keyCode) {
        case _keyCodes2.default.ESCAPE:
          e.preventDefault();
          _this.setState({
            isInputCancel: true,
            isTabEnter: false
          }, function () {
            target.blur();
          });
          break;
        case _keyCodes2.default.TAB:
        case _keyCodes2.default.ENTER:
          e.preventDefault();

          _this.setState({
            isTabEnter: true
          }, function () {
            target && target.blur();
          });
          break;
        default:
          break;
      }
    };

    _this.handleBlur = function (e) {
      var value = e.target.value;

      _this.setState({
        isInput: false
      }, function () {
        if (!_this.state.isInputCancel) {
          _this.props.onChange(value, _this.state.isTabEnter);
        } else {
          _this.props.onCancel();
        }
      });
    };

    _this.state = {
      isInput: false,
      isTabEnter: false
    };
    return _this;
  }

  Tag.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        editMode = _props.editMode,
        onClose = _props.onClose,
        value = _props.value,
        isNew = _props.isNew,
        rest = (0, _objectWithoutProperties3.default)(_props, ['className', 'editMode', 'onClose', 'value', 'isNew']);
    var isInput = this.state.isInput;

    var classes = (0, _classnames2.default)('tag-input', className);

    if (!editMode) {
      return _react2.default.createElement(
        'span',
        { className: 'badge' },
        value
      );
    }

    return _react2.default.createElement(
      'span',
      (0, _extends3.default)({ className: classes }, (0, _utils.omit)(rest, ['onChange', 'onCancel'])),
      isInput ? _react2.default.createElement('input', {
        className: 'input',
        ref: this.setInputRef,
        placeholder: isNew ? value : undefined,
        defaultValue: isNew ? undefined : value,
        onKeyDown: this.handleKeyDown,
        onBlur: this.handleBlur
      }) : _react2.default.createElement(
        'span',
        { className: 'badge badge--with-close-icon' },
        _react2.default.createElement(
          'span',
          { onClick: this.toInput },
          value
        ),
        _react2.default.createElement('i', { className: 'icon-close-thin', onClick: onClose })
      )
    );
  };

  return Tag;
}(_react.PureComponent);

Tag.displayName = 'AUITag';
Tag.propTypes = {
  isNew: _propTypes.bool,
  editMode: _propTypes.bool,
  className: _propTypes.string,
  value: _propTypes.string,
  onChange: _propTypes.func,
  onCancel: _propTypes.func,
  onClose: _propTypes.func
};
Tag.defaultProps = {
  editMode: false,
  onChange: _utils.noop,
  onCancel: _utils.noop,
  onClose: _utils.noop
};
exports.default = Tag;
module.exports = exports['default'];