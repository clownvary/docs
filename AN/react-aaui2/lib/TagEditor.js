'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _utils = require('./shared/utils');

var _Tag = require('./Tag');

var _Tag2 = _interopRequireDefault(_Tag);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var lastFocusedElement = null;

function trackLastFocusedElement() {
  lastFocusedElement = document.activeElement;
}

function stopPropagation(e) {
  e.stopPropagation();
}

var TagEditor = function (_PureComponent) {
  (0, _inherits3.default)(TagEditor, _PureComponent);

  function TagEditor() {
    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, TagEditor);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call.apply(_PureComponent, [this].concat(args))), _this), _this.setTagRef = function (tag) {
      _this.tag = tag;
    }, _this.handleClick = function () {
      if (lastFocusedElement && lastFocusedElement.tagName === 'INPUT') return;

      var d = _this.props.data;
      var data = d.concat({ text: 'your tag name', isNew: true });

      _this.props.onChange(data);
    }, _this.handleTagClose = function (ind) {
      return function () {
        var d = _this.props.data;
        var data = d.filter(function (item, index) {
          return index !== ind;
        });

        _this.props.onChange(data);
      };
    }, _this.handleTagChange = function (ind) {
      return function (text, isKeyFinished) {
        var data = _this.props.data;

        if (text === '') {
          data = data.filter(function (item, index) {
            return index !== ind;
          });
        } else {
          data = data.map(function (item, index) {
            if (index === ind) {
              return { text: text };
            }

            return item;
          });

          // If we the `keyCode` is `TAB` or `ENTER`, then add another new Tag
          if (isKeyFinished && ind === data.length - 1) {
            data = data.concat({ text: 'your tag name', isNew: true });
          }
        }

        _this.props.onChange(data);
      };
    }, _this.handleTabCancel = function (ind) {
      return function () {
        var data = _this.props.data;

        // Remove tag it it's new

        if (data[ind].isNew) {
          _this.props.onChange(data.filter(function (item, index) {
            return index !== ind;
          }));
        }
      };
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  TagEditor.prototype.componentDidUpdate = function componentDidUpdate(prevProps) {
    var lastTag = this.tag;

    this.props.data.length > prevProps.data.length && lastTag && lastTag.toInput();
  };

  TagEditor.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        errored = _props.errored,
        editMode = _props.editMode,
        className = _props.className,
        placeholder = _props.placeholder,
        data = _props.data,
        rest = (0, _objectWithoutProperties3.default)(_props, ['errored', 'editMode', 'className', 'placeholder', 'data']);

    var classes = (0, _classnames2.default)({
      tageditor: true,
      'is-errored': errored
    }, className);

    return _react2.default.createElement(
      'div',
      (0, _extends3.default)({
        className: classes
      }, (0, _utils.omit)(rest, ['onChange']), {
        onMouseDown: trackLastFocusedElement,
        onClick: editMode ? this.handleClick : _utils.noop
      }),
      data.map(function (_ref, i) {
        var text = _ref.text,
            isNew = _ref.isNew,
            tagRestProps = (0, _objectWithoutProperties3.default)(_ref, ['text', 'isNew']);
        return _react2.default.createElement(_Tag2.default, (0, _extends3.default)({
          key: i,
          ref: _this2.setTagRef,
          value: text,
          isNew: !!isNew,
          editMode: editMode,
          onChange: _this2.handleTagChange(i),
          onCancel: _this2.handleTabCancel(i),
          onClose: _this2.handleTagClose(i),
          onClick: stopPropagation
        }, tagRestProps));
      }),
      !data.length && placeholder
    );
  };

  return TagEditor;
}(_react.PureComponent);

TagEditor.displayName = 'AUITagEditor';
TagEditor.propTypes = {
  errored: _propTypes.bool,
  editMode: _propTypes.bool,
  className: _propTypes.string,
  onChange: _propTypes.func,

  placeholder: _propTypes.node,
  data: (0, _propTypes.oneOfType)([_propTypes.array, _propTypes.object]).isRequired
};
TagEditor.defaultProps = {
  errored: false,
  editMode: true,
  onChange: _utils.noop,
  data: []
};
exports.default = TagEditor;
module.exports = exports['default'];