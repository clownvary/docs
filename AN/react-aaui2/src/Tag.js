import React, { PureComponent } from 'react'
import { string, func, bool } from 'prop-types'
import classNames from 'classnames'

import { omit, noop } from './shared/utils'
import KEY_CODES from './shared/keyCodes'

export default class Tag extends PureComponent {
  static displayName = 'AUITag'

  static propTypes = {
    isNew: bool,
    editMode: bool,
    className: string,
    value: string,
    onChange: func,
    onCancel: func,
    onClose: func,
  }

  static defaultProps = {
    editMode: false,
    onChange: noop,
    onCancel: noop,
    onClose: noop,
  }

  constructor(props) {
    super(props)

    this.state = {
      isInput: false,
      isTabEnter: false,
    }
  }

  setInputRef = input => {
    this.input = input
  }

  toInput = e => {
    e && e.stopPropagation && e.stopPropagation()

    this.setState(
      {
        isInput: true,
        isInputCancel: false,
      },
      () => {
        const input = this.input

        if (input) {
          input.focus()
          input.select()
        }
      },
    )
  }

  handleKeyDown = e => {
    const target = e.target

    switch (e.keyCode) {
      case KEY_CODES.ESCAPE:
        e.preventDefault()
        this.setState(
          {
            isInputCancel: true,
            isTabEnter: false,
          },
          () => {
            target.blur()
          },
        )
        break
      case KEY_CODES.TAB:
      case KEY_CODES.ENTER:
        e.preventDefault()

        this.setState(
          {
            isTabEnter: true,
          },
          () => {
            target && target.blur()
          },
        )
        break
      default:
        break
    }
  }

  handleBlur = e => {
    const value = e.target.value

    this.setState(
      {
        isInput: false,
      },
      () => {
        if (!this.state.isInputCancel) {
          this.props.onChange(value, this.state.isTabEnter)
        } else {
          this.props.onCancel()
        }
      },
    )
  }

  render() {
    const { className, editMode, onClose, value, isNew, ...rest } = this.props
    const { isInput } = this.state
    const classes = classNames('tag-input', className)

    if (!editMode) {
      return <span className="badge">{value}</span>
    }

    return (
      <span className={classes} {...omit(rest, ['onChange', 'onCancel'])}>
        {isInput ? (
          <input
            className="input"
            ref={this.setInputRef}
            placeholder={isNew ? value : undefined}
            defaultValue={isNew ? undefined : value}
            onKeyDown={this.handleKeyDown}
            onBlur={this.handleBlur}
          />
        ) : (
          <span className="badge badge--with-close-icon">
            <span onClick={this.toInput}>{value}</span>
            <i className="icon-close-thin" onClick={onClose} />
          </span>
        )}
      </span>
    )
  }
}
