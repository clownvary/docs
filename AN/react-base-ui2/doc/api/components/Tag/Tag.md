## Tag 



### Prop Types
Property | Type | Default Value | Is Required | Description
:--- | :--- | :--- | :--- | :---
prefixCls|string|`${DefaultCSSPrefix}-tag`|false|&ensp;
className|string|&ensp;|false|class names which were applied to component container div.
size|enum:<br>[TagSize.SMALL, TagSize.MEDIUM, TagSize.LARGE, TagSize.EXTRA_LARGE]|TagSize.MEDIUM|false|determine the tag size.
type|enum:<br>[TagType.DEFAULT, TagType.PENDING, TagType.ERROR]|TagType.DEFAULT|false|determine the type of tag, including background-color and color.
closable|bool|&ensp;|false|determine the close icon renders of tag.
visible|bool|true|false|determine the hide status of tag.
onClose|func|&ensp;|false|the callback function which is triggered when clicking the close icon.