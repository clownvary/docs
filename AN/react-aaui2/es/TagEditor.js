import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent } from 'react';
import { bool, string, array, object, func, node, oneOfType } from 'prop-types';
import classNames from 'classnames';

import { omit, noop } from './shared/utils';
import Tag from './Tag';

var lastFocusedElement = null;

function trackLastFocusedElement() {
  lastFocusedElement = document.activeElement;
}

function stopPropagation(e) {
  e.stopPropagation();
}

var TagEditor = function (_PureComponent) {
  _inherits(TagEditor, _PureComponent);

  function TagEditor() {
    var _temp, _this, _ret;

    _classCallCheck(this, TagEditor);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _PureComponent.call.apply(_PureComponent, [this].concat(args))), _this), _this.setTagRef = function (tag) {
      _this.tag = tag;
    }, _this.handleClick = function () {
      if (lastFocusedElement && lastFocusedElement.tagName === 'INPUT') return;

      var d = _this.props.data;
      var data = d.concat({ text: 'your tag name', isNew: true });

      _this.props.onChange(data);
    }, _this.handleTagClose = function (ind) {
      return function () {
        var d = _this.props.data;
        var data = d.filter(function (item, index) {
          return index !== ind;
        });

        _this.props.onChange(data);
      };
    }, _this.handleTagChange = function (ind) {
      return function (text, isKeyFinished) {
        var data = _this.props.data;

        if (text === '') {
          data = data.filter(function (item, index) {
            return index !== ind;
          });
        } else {
          data = data.map(function (item, index) {
            if (index === ind) {
              return { text: text };
            }

            return item;
          });

          // If we the `keyCode` is `TAB` or `ENTER`, then add another new Tag
          if (isKeyFinished && ind === data.length - 1) {
            data = data.concat({ text: 'your tag name', isNew: true });
          }
        }

        _this.props.onChange(data);
      };
    }, _this.handleTabCancel = function (ind) {
      return function () {
        var data = _this.props.data;

        // Remove tag it it's new

        if (data[ind].isNew) {
          _this.props.onChange(data.filter(function (item, index) {
            return index !== ind;
          }));
        }
      };
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  TagEditor.prototype.componentDidUpdate = function componentDidUpdate(prevProps) {
    var lastTag = this.tag;

    this.props.data.length > prevProps.data.length && lastTag && lastTag.toInput();
  };

  TagEditor.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        errored = _props.errored,
        editMode = _props.editMode,
        className = _props.className,
        placeholder = _props.placeholder,
        data = _props.data,
        rest = _objectWithoutProperties(_props, ['errored', 'editMode', 'className', 'placeholder', 'data']);

    var classes = classNames({
      tageditor: true,
      'is-errored': errored
    }, className);

    return React.createElement(
      'div',
      _extends({
        className: classes
      }, omit(rest, ['onChange']), {
        onMouseDown: trackLastFocusedElement,
        onClick: editMode ? this.handleClick : noop
      }),
      data.map(function (_ref, i) {
        var text = _ref.text,
            isNew = _ref.isNew,
            tagRestProps = _objectWithoutProperties(_ref, ['text', 'isNew']);

        return React.createElement(Tag, _extends({
          key: i,
          ref: _this2.setTagRef,
          value: text,
          isNew: !!isNew,
          editMode: editMode,
          onChange: _this2.handleTagChange(i),
          onCancel: _this2.handleTabCancel(i),
          onClose: _this2.handleTagClose(i),
          onClick: stopPropagation
        }, tagRestProps));
      }),
      !data.length && placeholder
    );
  };

  return TagEditor;
}(PureComponent);

TagEditor.displayName = 'AUITagEditor';
TagEditor.propTypes = {
  errored: bool,
  editMode: bool,
  className: string,
  onChange: func,

  placeholder: node,
  data: oneOfType([array, object]).isRequired
};
TagEditor.defaultProps = {
  errored: false,
  editMode: true,
  onChange: noop,
  data: []
};
export default TagEditor;