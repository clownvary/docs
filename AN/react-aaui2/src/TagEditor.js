import React, { PureComponent } from 'react'
import { bool, string, array, object, func, node, oneOfType } from 'prop-types'
import classNames from 'classnames'

import { omit, noop } from './shared/utils'
import Tag from './Tag'

let lastFocusedElement = null

function trackLastFocusedElement() {
  lastFocusedElement = document.activeElement
}

function stopPropagation(e) {
  e.stopPropagation()
}

export default class TagEditor extends PureComponent {
  static displayName = 'AUITagEditor'

  static propTypes = {
    errored: bool,
    editMode: bool,
    className: string,
    onChange: func,

    placeholder: node,
    data: oneOfType([array, object]).isRequired,
  }

  static defaultProps = {
    errored: false,
    editMode: true,
    onChange: noop,
    data: [],
  }

  componentDidUpdate(prevProps) {
    const lastTag = this.tag

    this.props.data.length > prevProps.data.length &&
      lastTag &&
      lastTag.toInput()
  }

  setTagRef = tag => {
    this.tag = tag
  }

  handleClick = () => {
    if (lastFocusedElement && lastFocusedElement.tagName === 'INPUT') return

    const d = this.props.data
    const data = d.concat({ text: 'your tag name', isNew: true })

    this.props.onChange(data)
  }

  handleTagClose = ind => () => {
    const d = this.props.data
    const data = d.filter((item, index) => index !== ind)

    this.props.onChange(data)
  }

  handleTagChange = ind => (text, isKeyFinished) => {
    let data = this.props.data

    if (text === '') {
      data = data.filter((item, index) => index !== ind)
    } else {
      data = data.map((item, index) => {
        if (index === ind) {
          return { text }
        }

        return item
      })

      // If we the `keyCode` is `TAB` or `ENTER`, then add another new Tag
      if (isKeyFinished && ind === data.length - 1) {
        data = data.concat({ text: 'your tag name', isNew: true })
      }
    }

    this.props.onChange(data)
  }

  handleTabCancel = ind => () => {
    const { data } = this.props

    // Remove tag it it's new
    if (data[ind].isNew) {
      this.props.onChange(data.filter((item, index) => index !== ind))
    }
  }

  render() {
    const {
      errored,
      editMode,
      className,
      placeholder,
      data,
      ...rest
    } = this.props
    const classes = classNames(
      {
        tageditor: true,
        'is-errored': errored,
      },
      className,
    )

    return (
      <div
        className={classes}
        {...omit(rest, ['onChange'])}
        onMouseDown={trackLastFocusedElement}
        onClick={editMode ? this.handleClick : noop}
      >
        {data.map(({ text, isNew, ...tagRestProps }, i) => (
          <Tag
            key={i}
            ref={this.setTagRef}
            value={text}
            isNew={!!isNew}
            editMode={editMode}
            onChange={this.handleTagChange(i)}
            onCancel={this.handleTabCancel(i)}
            onClose={this.handleTagClose(i)}
            onClick={stopPropagation}
            {...tagRestProps}
          />
        ))}
        {!data.length && placeholder}
      </div>
    )
  }
}
