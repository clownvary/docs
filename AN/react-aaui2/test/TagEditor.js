import React from 'react'
import renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'

import TagEditor from 'TagEditor'
import Tag from 'Tag'

const data = [
  {
    text: 'first',
  },
]

beforeEach(() => {
  document.activeElement && document.activeElement.blur()
})

it('should render without errors', () => {
  const wrapper = shallow(<TagEditor />)

  expect(wrapper).toBeTruthy()
})

it("should render right thing and don't change unexpected", () => {
  const tree = renderer.create(<TagEditor data={data} />).toJSON()

  expect(tree).toMatchSnapshot()
})

it('should has `.tageditor.is-errored` when setting errored as true', () => {
  const wrapper = shallow(<TagEditor errored />)

  expect(wrapper.hasClass('is-errored')).toBe(true)
})

it('should call `handleClick` right', () => {
  const onChangeMock = jest.fn()
  const wrapper = shallow(<TagEditor data={data} onChange={onChangeMock} />)
  const tagEditor = wrapper.instance()

  tagEditor.handleClick()

  expect(onChangeMock).toBeCalled()
  expect(onChangeMock).toBeCalledWith([
    {
      text: 'first',
    },
    {
      isNew: true,
      text: 'your tag name',
    },
  ])
})

it('should call `handleTagClose` right', () => {
  const onChangeMock = jest.fn()
  const wrapper = shallow(<TagEditor data={data} onChange={onChangeMock} />)
  const tagEditor = wrapper.instance()

  tagEditor.handleTagClose(0)()

  expect(onChangeMock).toBeCalled()
  expect(onChangeMock).toBeCalledWith([])
})

it('should call `handleTabCancel` right', () => {
  const onChangeMock = jest.fn()

  data[0].isNew = true
  const wrapper = shallow(<TagEditor data={data} onChange={onChangeMock} />)
  const tagEditor = wrapper.instance()

  tagEditor.handleTabCancel(0)()

  expect(onChangeMock).toBeCalled()
  expect(onChangeMock).toBeCalledWith([])
})

it('should call `handleTagChange` right', () => {
  const onChangeMock = jest.fn()
  const wrapper = shallow(<TagEditor data={data} onChange={onChangeMock} />)
  const tagEditor = wrapper.instance()

  tagEditor.handleTagChange(0)('Updated tag text')

  expect(onChangeMock).toBeCalled()
  expect(onChangeMock).toBeCalledWith([
    {
      text: 'Updated tag text',
    },
  ])

  tagEditor.handleTagChange(0)('Updated tag text', true)

  expect(onChangeMock).toBeCalled()
  expect(onChangeMock).toBeCalledWith([
    {
      text: 'Updated tag text',
    },
    {
      isNew: true,
      text: 'your tag name',
    },
  ])
})

it('should call componentDidUpdate right', () => {
  const onChangeMock = jest.fn()
  const wrapper = mount(<TagEditor data={data} onChange={onChangeMock} />)
  const tagEditor = wrapper.instance()
  wrapper.setProps({
    data: data.concat({
      text: 'text',
    }),
  })
  wrapper.find('.tageditor').simulate('mouseDown')
  tagEditor.handleTagChange(0)('', true)
  wrapper
    .find('.tageditor')
    .childAt(0)
    .simulate('click')
  wrapper.setProps({
    data: data.concat([
      {
        text: 'text',
      },
      {
        text: 'text2',
      },
    ]),
  })
  expect(onChangeMock).toBeCalled()
})

it('should support custom props for every tag', () => {
  const customProps = {
    className: 'errored',
    a: 1,
    b: 2,
  }
  const customData = [
    {
      text: 'first',
      ...customProps,
    },
  ]
  const Component = <TagEditor data={customData} />
  const wrapper = mount(Component)
  const tree = renderer.create(Component).toJSON()

  expect(wrapper.find(Tag).props()).toMatchObject(
    expect.objectContaining(customProps),
  )
  expect(tree).toMatchSnapshot()
})

it('should render `placeholder` when `data` is empty', () => {
  const Component = (
    <TagEditor
      placeholder={<p className="strong">Your placeholder is here!</p>}
    />
  )

  expect(renderer.create(Component).toJSON()).toMatchSnapshot()
})

it('should could disable the editing with `editMode` as false', () => {
  const onChange = jest.fn()
  const Component = <TagEditor onChange={onChange} editMode={false} />
  const wrapper = mount(Component)

  expect(renderer.create(Component).toJSON()).toMatchSnapshot()

  wrapper.find('.tageditor').simulate('mouseDown')
  wrapper.find('.tageditor').simulate('click')
  expect(onChange).not.toBeCalled()
})
