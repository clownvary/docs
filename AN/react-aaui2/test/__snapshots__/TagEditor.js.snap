// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should could disable the editing with \`editMode\` as false 1`] = `
<div
  className="tageditor"
  onClick={[Function]}
  onMouseDown={[Function]}
/>
`;

exports[`should render \`placeholder\` when \`data\` is empty 1`] = `
<div
  className="tageditor"
  onClick={[Function]}
  onMouseDown={[Function]}
>
  <p
    className="strong"
  >
    Your placeholder is here!
  </p>
</div>
`;

exports[`should render right thing and don't change unexpected 1`] = `
<div
  className="tageditor"
  onClick={[Function]}
  onMouseDown={[Function]}
>
  <span
    className="tag-input"
    onClick={[Function]}
  >
    <span
      className="badge badge--with-close-icon"
    >
      <span
        onClick={[Function]}
      >
        first
      </span>
      <i
        className="icon-close-thin"
        onClick={[Function]}
      />
    </span>
  </span>
</div>
`;

exports[`should support custom props for every tag 1`] = `
<div
  className="tageditor"
  onClick={[Function]}
  onMouseDown={[Function]}
>
  <span
    a={1}
    b={2}
    className="tag-input errored"
    onClick={[Function]}
  >
    <span
      className="badge badge--with-close-icon"
    >
      <span
        onClick={[Function]}
      >
        first
      </span>
      <i
        className="icon-close-thin"
        onClick={[Function]}
      />
    </span>
  </span>
</div>
`;
